### PR TITLE
Add rp2040 zero

### PR DESCRIPTION
**Pull request by glebsexy**
## Board/Shield Check-list

- [x] This board/shield is tested working on real hardware
- [x] Definitions follow the general style of other shields/boards upstream ([Reference](https://zmk.dev/docs/development/new-shield))
- [x] `.zmk.yml` metadata file added
- [x] Proper Copyright + License headers added to applicable files (Generally, we stick to "The ZMK Contributors" for copyrights to help avoid churn when files get edited)
- [x] General consistent formatting of DeviceTree files
- [ ] n/a Keymaps do not use deprecated key defines (Check using the [upgrader tool](https://zmk.dev/docs/codes/keymap-upgrader))
- [ ] n/a `&pro_micro` used in favor of `&pro_micro_d/a` if applicable
- [ ] n/a If split, no name added for the right/peripheral half
- [x] Kconfig.defconfig file correctly wraps _all_ configuration in conditional on the shield symbol
- [ ] n/a `.conf` file has optional extra features commented out
- [x] Keyboard/PCB is part of a shipped group buy or is generally available in stock to purchase (OSH/personal projects without general availability should create a zmk-config repo instead)

### CODE REVIEW

File: app/boards/arm/waveshare_rp2040_zero/waveshare_rp2040_zero.dts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+#include <rpi_pico/rp2040.dtsi>
+#include "waveshare_rp2040_zero-pinctrl.dtsi"
+#include <freq.h>
+
+/ {
+	chosen {
+		zephyr,sram = &sram0;
+		zephyr,flash = &flash0;
+		zephyr,console = &uart0;
```
Comment (petejohanson) [inline app/boards/arm/waveshare_rp2040_zero/waveshare_rp2040_zero.dts:17]: For boards added to ZMK directly, we usually set the console to be the CDC ACM node added to the usb node. See https://github.com/zmkfirmware/zmk/blob/main/app/boards/arm/nice_nano/nice_nano.dtsi#L64 for an example node.

File: app/boards/arm/waveshare_rp2040_zero/waveshare_rp2040_zero.dts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+		storage_partition: partition@7fbe00 {
+			label = "storage";
+			reg = <0x007fbe00 DT_SIZE_K(16)>;
+		};
+	};
+};
+
+&uart1 {
+	current-speed = <115200>;
+	status = "okay";
```
Comment (petejohanson) [inline app/boards/arm/waveshare_rp2040_zero/waveshare_rp2040_zero.dts:63]: For ZMK boards, don't enable the UART, SPI, etc by default. This is a bit different from the upstream Zephyr board approach.

File: app/boards/arm/waveshare_rp2040_zero/waveshare_rp2040_zero.zmk.yml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+identifier: waveshare_rp2040_zero
```
Comment (petejohanson) [inline app/boards/arm/waveshare_rp2040_zero/waveshare_rp2040_zero.zmk.yml:1]: This `.zmk.yml` file format is *not* the same as the Zephyr format. See existing ZMK boards for examples of the metadata we use.

**Review by petejohanson — CHANGES_REQUESTED** [review]
Thanks for the PR! A few items that need addressing. Also check out the pre-commit check failure for issues there.

File: app/boards/arm/waveshare_rp2040_zero/waveshare_rp2040_zero_defconfig [hunk 4]
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2023 The ZMK Contributors
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_SOC_SERIES_RP2XXX=y
+CONFIG_SOC_RP2040=y
+CONFIG_BOARD_WAVESHARE_RP2040_ZERO=y
+
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=240000000
```
Comment (petejohanson) [inline app/boards/arm/waveshare_rp2040_zero/waveshare_rp2040_zero_defconfig:8]:
```suggestion
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=125000000
```

I had to fix this in our other boards recently, they had the wrong sys clock frequency set.

**Comment by xudongzheng** [comment]
@petejohanson I noticed that ZMK's Zephyr fork has quite a few custom RP2040 boards defined there https://github.com/zmkfirmware/zephyr/tree/v3.2.0%2Bzmk-fixes/boards/arm. Is there a rule of thumb for which boards should go in upstream Zephyr, the ZMK Zephyr fork, vs the ZMK boards directory?

**Comment by petejohanson** [comment]
> @petejohanson I noticed that ZMK's Zephyr fork has quite a few custom RP2040 boards defined there https://github.com/zmkfirmware/zephyr/tree/v3.2.0%2Bzmk-fixes/boards/arm. Is there a rule of thumb for which boards should go in upstream Zephyr, the ZMK Zephyr fork, vs the ZMK boards directory?

Anything that I'd like to upstream, given time, I try to put in our Zephyr fork. That usually means anything more popular/general purpose.

The Zero is theoretically generic enough we could choose to put there as well.

**Comment by petejohanson** [comment]
Closing due to inactivity. Feel free to reopen whenever you're back to this.

**Pull request closed without merging**